### ISSUE TITLE
Issue submitting pydoop jobs (NullPointerException)

### ISSUE THREAD

**Issue by dbastide** [crs4/pydoop#181]
Hi,

I installed Pydoop on a VirtualBox machine based on Centos7 with Hadoop 2.6.3, by running the command: ```pip install pydoop```
Then I downloaded Pydoop code from github to run "all_tests.py".
All tests are successful except the first one: 
```
test_conf_file (test_submit.TestAppSubmit) ... FAIL
Traceback (most recent call last):
  File "/root/pydoop_code/pydoop/test/app/test_submit.py", line 124, in test_conf_file
    self.assertEqual(len(unknown), 0)
AssertionError: 1 != 0
```
I don't know if this error for this particular test could be linked to my issues below.

I can successfully run a Hadoop wordcount exemple from a jar file with:
```hadoop jar /root/wc.jar hadoop.WordCount /user/root/input/smallText.txt /user/root/output```
so it seems my hadoop installation is working.
However, Pydoop jobs fail: I tried to launch the wordcount examples by running ./run_wc_minimal or ./run_wc_minimal_new_api but all my attempts fail with the same NullPointerException.

Here is what I get in the console:
```
[root@localhost wordcount]# ./run_wc_minimal_new_api
INFO:main:copying input data to HDFS
INFO:main:running MapReduce application
DEPRECATED: Use of this script to execute mapred command is deprecated.
Instead use the mapred command for it.

16/01/14 10:05:54 INFO Configuration.deprecation: session.id is deprecated. Instead, use dfs.metrics.session-id
16/01/14 10:05:54 INFO jvm.JvmMetrics: Initializing JVM Metrics with processName=JobTracker, sessionId=
16/01/14 10:05:54 INFO jvm.JvmMetrics: Cannot initialize JVM Metrics with processName=JobTracker, sessionId= - already initialized
16/01/14 10:05:55 WARN mapreduce.JobResourceUploader: No job jar file set.  User classes may not be found. See Job or Job#setJar(String).
16/01/14 10:05:55 INFO mapred.FileInputFormat: Total input paths to process : 1
16/01/14 10:05:55 INFO mapreduce.JobSubmitter: number of splits:1
16/01/14 10:05:55 INFO Configuration.deprecation: hadoop.pipes.java.recordreader is deprecated. Instead, use mapreduce.pipes.isjavarecordreader
16/01/14 10:05:55 INFO Configuration.deprecation: mapred.reduce.tasks is deprecated. Instead, use mapreduce.job.reduces
16/01/14 10:05:55 INFO Configuration.deprecation: mapred.job.name is deprecated. Instead, use mapreduce.job.name
16/01/14 10:05:55 INFO Configuration.deprecation: hadoop.pipes.java.recordwriter is deprecated. Instead, use mapreduce.pipes.isjavarecordwriter
16/01/14 10:05:55 INFO Configuration.deprecation: mapred.map.tasks is deprecated. Instead, use mapreduce.job.maps
16/01/14 10:05:56 INFO mapreduce.JobSubmitter: Submitting tokens for job: job_local1111854526_0001
16/01/14 10:05:57 INFO mapred.LocalDistributedCacheManager: Creating symlink: /tmp/hadoop-root/mapred/local/1452762356710/pydoop_3b473d6545034f059823ec2147c368aa <- /root/pydoop_code/pydoop/examples/wordcount/pydoop_3b473d6545034f059823ec2147c368aa
16/01/14 10:05:57 INFO mapred.LocalDistributedCacheManager: Localized hdfs://localhost:9000/user/root/pydoop_cfde31830cf64382b0b6fd7e46e1bb0a/pydoop_3b473d6545034f059823ec2147c368aa as file:/tmp/hadoop-root/mapred/local/1452762356710/pydoop_3b473d6545034f059823ec2147c368aa
16/01/14 10:05:57 INFO mapreduce.Job: The url to track the job: http://localhost:8080/
16/01/14 10:05:57 INFO mapreduce.Job: Running job: job_local1111854526_0001
16/01/14 10:05:57 INFO mapred.LocalJobRunner: OutputCommitter set in config null
16/01/14 10:05:57 INFO mapred.LocalJobRunner: OutputCommitter is org.apache.hadoop.mapred.FileOutputCommitter
16/01/14 10:05:57 INFO mapred.LocalJobRunner: Waiting for map tasks
16/01/14 10:05:57 INFO mapred.LocalJobRunner: Starting task: attempt_local1111854526_0001_m_000000_0
16/01/14 10:05:57 INFO mapred.Task:  Using ResourceCalculatorProcessTree : [ ]
16/01/14 10:05:57 INFO mapred.MapTask: Processing split: hdfs://localhost:9000/user/root/pydoop_cfde31830cf64382b0b6fd7e46e1bb0a/input/alice.txt:0+167529
16/01/14 10:05:57 INFO mapred.MapTask: numReduceTasks: 2
16/01/14 10:05:57 INFO mapred.MapTask: (EQUATOR) 0 kvi 26214396(104857584)
16/01/14 10:05:57 INFO mapred.MapTask: mapreduce.task.io.sort.mb: 100
16/01/14 10:05:57 INFO mapred.MapTask: soft limit at 83886080
16/01/14 10:05:57 INFO mapred.MapTask: bufstart = 0; bufvoid = 104857600
16/01/14 10:05:57 INFO mapred.MapTask: kvstart = 26214396; length = 6553600
16/01/14 10:05:57 INFO mapred.MapTask: Map output collector class = org.apache.hadoop.mapred.MapTask$MapOutputBuffer
16/01/14 10:05:57 INFO mapred.LocalJobRunner: map task executor complete.
16/01/14 10:05:57 WARN mapred.LocalJobRunner: job_local1111854526_0001
java.lang.Exception: java.lang.NullPointerException
        at org.apache.hadoop.mapred.LocalJobRunner$Job.runTasks(LocalJobRunner.java:462)
        at org.apache.hadoop.mapred.LocalJobRunner$Job.run(LocalJobRunner.java:522)
Caused by: java.lang.NullPointerException
        at org.apache.hadoop.mapred.pipes.Application.<init>(Application.java:104)
        at org.apache.hadoop.mapred.pipes.PipesMapRunner.run(PipesMapRunner.java:69)
        at org.apache.hadoop.mapred.MapTask.runOldMapper(MapTask.java:450)
        at org.apache.hadoop.mapred.MapTask.run(MapTask.java:343)
        at org.apache.hadoop.mapred.LocalJobRunner$Job$MapTaskRunnable.run(LocalJobRunner.java:243)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:471)
        at java.util.concurrent.FutureTask.run(FutureTask.java:262)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
        at java.lang.Thread.run(Thread.java:744)
16/01/14 10:05:58 INFO mapreduce.Job: Job job_local1111854526_0001 running in uber mode : false
16/01/14 10:05:58 INFO mapreduce.Job:  map 0% reduce 0%
16/01/14 10:05:58 INFO mapreduce.Job: Job job_local1111854526_0001 failed with state FAILED due to: NA
16/01/14 10:05:58 INFO mapreduce.Job: Counters: 0
Exception in thread "main" java.io.IOException: Job failed!
        at org.apache.hadoop.mapred.JobClient.runJob(JobClient.java:836)
        at org.apache.hadoop.mapred.pipes.Submitter.runJob(Submitter.java:264)
        at org.apache.hadoop.mapred.pipes.Submitter.run(Submitter.java:503)
        at org.apache.hadoop.mapred.pipes.Submitter.main(Submitter.java:518)
Traceback (most recent call last):
  File "run_wc.py", line 79, in <module>
    main(sys.argv[1:])
  File "run_wc.py", line 71, in main
    runner.run(properties=CONF, hadoop_conf_dir=HADOOP_CONF_DIR, logger=logger)
  File "/usr/lib/python2.7/site-packages/pydoop/hadut.py", line 495, in run
    run_pipes(self.exe, self.input, self.output, **kwargs)
  File "/usr/lib/python2.7/site-packages/pydoop/hadut.py", line 365, in run_pipes
    keep_streams=keep_streams)
  File "/usr/lib/python2.7/site-packages/pydoop/hadut.py", line 167, in run_cmd
    raise RunCmdError(p.returncode, ' '.join(_args), error)
pydoop.hadut.RunCmdError: command exited with 1 status
```
If I look at Application.java:104, I see
```
 99    //Add token to the environment if security is enabled
 100  Token<JobTokenIdentifier> jobToken = TokenCache.getJobToken(conf
 101       .getCredentials());
 102  // This password is used as shared secret key between this application and
 103  // child pipes process
 104   byte[]  password = jobToken.getPassword();
```

Do you have any idea of what could be wrong? I did not configure anything about security. What should I do?
Thanks in advance!
David

**Comment by ilveroluca** [comment]
Hi David.  I'm not sure what the problem is.  Maybe you've found a bug.  

What flavour of Hadoop are you using?  Would you be comfortable doing some debugging with us?

**Comment by dbastide** [comment]
Hi, 

Thank you for your answer. Actually, I fixed my issue! The first unit test still fails but all the others pass, and I can now launch the examples. I didn't post my answer sooner because I wanted to be sure everything was really ok.

I was using a standalone hadoop, with the default configuration.
After updating the configuration files (mapred-site.xml and yarn-site.xml updates were probably the main cause for the fix) and restarting the services, the error disappeared.
Here is the configuration I have now:
core-site.xml:
```
<configuration>
    <property>
        <name>fs.defaultFS</name>
        <value>hdfs://localhost:9000</value>
    </property>
</configuration>
```
hdfs-site.xml:
```
<configuration>
    <property>
        <name>dfs.replication</name>
        <value>1</value>
    </property>
</configuration>
```
mapred-site.xml:
```
<configuration>
        <property>
            <name>mapreduce.framework.name</name>
            <value>yarn</value>
        </property>
</configuration>
```
yarn-site.xml:
```
<configuration>
    <property>
        <name>yarn.nodemanager.aux-services</name>
        <value>mapreduce_shuffle</value>
    </property>
</configuration>
```

**Comment by ilveroluca** [comment]
Thanks for your feedback.  Do you still have the configuration that generated the error?  I'd be interested in seeing it.

**Comment by dbastide** [comment]
I tried a lot of things so I'm not 100% sure, but I think when I posted the question I had this configuration:

core-site.xml:
```
<configuration>
    <property>
        <name>fs.defaultFS</name>
        <value>hdfs://localhost:9000</value>
    </property>
</configuration>
```
hdfs-site.xml:
```
<configuration>
    <property>
        <name>dfs.replication</name>
        <value>1</value>
    </property>
</configuration>
```
mapred-site.xml:
```
<configuration>
</configuration>
```
yarn-site.xml:
```
<configuration>
</configuration>
```

**Comment by ilveroluca** [comment]
Thanks.

**Comment by ilveroluca** [comment]
I'm re-opening this issue since I've figured out how to reproduce it.

It seems to be related to the `mapreduce.framework.name` being used.  The `NullPointerException` is raised if I launch an application with the `local` framework; that's the default value of the property, so it's used when the property isn't set.  If I set the property to the value `yarn` (like @dbastide did above) the problem goes away.

**Comment by mdrio** [comment]
Hi, local mode is not supported, since Hadoop pipes can run only in distributed or pseudo distributed mode. We will add an error message when using local mode.

**Comment by simleo** [comment]
2e765cebfafe6595edaafa49f7ce91f22b387ee5 is the bad guy according to `git bisect`

**Comment by elzaggo** [comment]
Closed by #194.
